Simulate in physical coordinates

daniel-goldstein opened the pull request:
Simulations now are computed in physical coordinates. The Fenwick tree used to track and sample segments measures cumulative recombination mass. Supports both discrete and continuous physical space (using the `discrete` flag on `RecombinationMap`).

jeromekelleher commented on lib/recomb_map.c:
```diff
@@ -214,48 +194,39 @@ recomb_map_phys_to_discrete_genetic(recomb_map_t *self, double x, uint32_t *locu
 double
 recomb_map_genetic_to_phys(recomb_map_t *self, double genetic_x)
 {
```
Ditto here, we want this function as the higher level APIs that we're supporting still provide it.

jeromekelleher commented on lib/recomb_map.c:
```diff
@@ -214,48 +194,39 @@ recomb_map_phys_to_discrete_genetic(recomb_map_t *self, double x, uint32_t *locu
+double
+recomb_map_position_to_recomb_mass(recomb_map_t *self, double pos)
+{
+    if (pos == self->positions[0]) return 0;
```
Always use braces:
```c
if (pos == self->positions[0]) {
    return 0;
}
```

jeromekelleher commented on lib/recomb_map.c:
```diff
@@ -214,48 +194,39 @@ recomb_map_phys_to_discrete_genetic(recomb_map_t *self, double x, uint32_t *locu
-    }
-    return ret;
+double
+recomb_map_position_to_recomb_mass(recomb_map_t *self, double pos)
```
Rename to ``recomb_map_position_to_mass()``

A quick comment describing what it does would help too.

jeromekelleher commented on lib/recomb_map.c:
```diff
@@ -214,48 +194,39 @@ recomb_map_phys_to_discrete_genetic(recomb_map_t *self, double x, uint32_t *locu
+    if (pos == self->positions[0]) return 0;
+    if (pos >= self->positions[self->size - 1]) return self->cumulative[self->size - 1];
+
+    size_t right = msp_binary_interval_search(pos, self->positions, self->size);
```
``left`` and ``right`` arent great names. What about 
```c
double offset;
size_t index = msp_binary_interval_search(pos, self->positions, self->size);
assert(index != 0)
index--;
offset = ..
```

jeromekelleher commented on lib/recomb_map.c:
```diff
@@ -214,48 +194,39 @@ recomb_map_phys_to_discrete_genetic(recomb_map_t *self, double x, uint32_t *locu
+}
+
+static double
+recomb_map_recomb_mass_to_position(recomb_map_t *self, double recomb_mass)
```
rename ``recomb_map_mass_to_position``

daniel-goldstein commented on lib/fenwick.c:
```diff
@@ -49,12 +49,12 @@ fenwick_alloc_buffers(fenwick_t *self)
     self->tree = NULL;
     self->values = NULL;
-    self->tree = calloc((1 + self->size), sizeof(int64_t));
+    self->tree = calloc((1 + self->size), sizeof(self->tree));
```
yup

daniel-goldstein commented on lib/recomb_map.c:
```diff
@@ -166,30 +168,8 @@ recomb_map_get_num_loci(recomb_map_t *self)
 double
 recomb_map_phys_to_genetic(recomb_map_t *self, double x)
```
I'll keep these around and remove unnecessary calls to them instead. I thought we were going to remove them entirely but either way this is cleaner for now.

jeromekelleher replied:
Sounds good. Making them no-ops was a good quick hack to see if this all worked, but we want to start doing things properly now.

daniel-goldstein commented:
Addressed comments. I'm a bit wary of the effect on GC and other events although I _think_ I removed only the necessary translations of genetic -> phys. I believe GC should still operate fine in genetic space. If things look ok I'll squash.

daniel-goldstein commented on msprime/simulations.py:
```diff
@@ -967,17 +988,19 @@ def get_per_locus_recombination_rate(self):
     def get_size(self):
         return self._ll_recombination_map.get_size()
 
+    @deprecated(reason="Use sequence_length instead")
```
remove

daniel-goldstein commented on tests/test_lowlevel.py:
```diff
@@ -36,15 +36,12 @@
 
 
-def uniform_recombination_map(num_loci=1, rate=0, L=None):
+def uniform_recombination_map(L=1, rate=0):
```
skip failing num_loci tests

jeromekelleher commented on msprime/simulations.py:
```diff
@@ -175,6 +175,7 @@ def simulator_factory(
         length=None,
         recombination_rate=None,
         recombination_map=None,
+        finite_recombination_sites=False,
```
Let's remove this parameter for now and have a think about what we want the API to be.

jeromekelleher commented on msprime/simulations.py:
```diff
@@ -819,57 +825,77 @@ class RecombinationMap(object):
+            if num_loci == positions[-1]:
+                warnings.warn(
+                        "num_loci is deprecated and should not be used. "
+                        "The default behavior is that the number of loci "
```
Remove the second sentence.

jeromekelleher commented on msprime/simulations.py:
```diff
@@ -819,57 +825,77 @@ class RecombinationMap(object):
-        simulation, and is by default large enough to be effectively be
-        a continuous model.
+        ``discrete`` controls whether recombination sites can occur only on integer
+        positions or can take continuous values. The optional ``num_loci`` is
```
It's not so much deprecated as removed. "The legacy ``num_loci`` option is no longer supported and should not be used".

jeromekelleher commented on msprime/simulations.py:
```diff
@@ -819,57 +825,77 @@ class RecombinationMap(object):
-            >>> recomb_map = RecombinationMap.uniform_map(m, rate, num_loci=m)
+            >>> recomb_map = RecombinationMap.uniform_map(m, rate, discrete=True)
+
+        .. warning::
```
I don't think we need a warning here, can delete.

jeromekelleher commented on msprime/simulations.py:
```diff
@@ -819,57 +825,77 @@ class RecombinationMap(object):
         :param float rate: The rate of recombination per unit of sequence length
             along this chromosome.
-        :param int num_loci: The number of discrete loci in the underlying
+        :param int num_loci: **This parameter is deprecated**.
```
this parameter is no longer supported

jeromekelleher replied:
Delete the old description -- we just want to tell the user that this positional parameter is there, and that it shouldn't be used.

jeromekelleher commented:
This is getting pretty close and could use some more eyes on it. Most things work now, except for recombination in the DTWF which we'll have to rethink a bit. (I think we should probably do this as a follow up PR). The genic selection code seems to have been roughed up a bit too.

@petrelharp, @andrewkern, @DomNelson, @fbaumdicker, any chance you could take a look through please?

daniel-goldstein commented:
Updated DTWF so that breakpoints are sampled using a rate equivalent to the average rate of recombination from [x->left, sequence_length). Tests for DTWF are now passing and no longer skipped

jeromekelleher commented:
Awesome, thanks @daniel-goldstein. Let's clear up any remaining skips if we can, and get ready to merge.

Has the algorithms.py file been updated to use the new logic?

daniel-goldstein commented:
> Has the algorithms.py file been updated to use the new logic?

it should be mostly there because it doesn't have much of the typing issues.  Will make sure it's all synced.

daniel-goldstein commented:
No new skips now. `algorithms.py` now reflects the backend logic

jeromekelleher commented on algorithms.py:
```diff
@@ -1705,27 +1802,28 @@ def verify(self):
+                                            u.left, u.right)
                         right = u.right
+                        if self.model != 'dtwf' and self.model != 'wf_ped':
+                            assert abs(s - self.L[l].get_frequency(u.index)) < 1e-6
```
Use math.isclose() here.

jeromekelleher replied:
And elsewhere

daniel-goldstein commented:
Yup, everything seems to check out fine for `ms` and `mshot` comparisons. Here's the breakpoint distribution on the latest commit for `mshot-high-recomb-variance`
![recomb map coalescent_breakpoints](https://user-images.githubusercontent.com/24440116/73689122-6b3ec980-46c5-11ea-8886-5a0d8551694e.png)

jeromekelleher commented:
Can you run some of the dtwf checks as well please? Once these pass we're good to merge I think.

daniel-goldstein commented:
> Can you run some of the dtwf checks as well please? Once these pass we're good to merge I think.

They all seem pretty consistent. Here's the `num_trees` and `tmrca_mean` for

`dtwf_vs_coalescent_long_region`:
![num_trees](https://user-images.githubusercontent.com/24440116/73735192-ca88f200-4736-11ea-8ae7-65508f078008.png)
![tmrca_mean](https://user-images.githubusercontent.com/24440116/73735194-ca88f200-4736-11ea-856f-eb3501441724.png)


`dtwf_vs_coalescent_3_pops_asymm_mig`:
![num_trees](https://user-images.githubusercontent.com/24440116/73735126-af1de700-4736-11ea-83d9-eff656a81b7e.png)
![tmrca_mean](https://user-images.githubusercontent.com/24440116/73735127-afb67d80-4736-11ea-984c-d12c33295618.png)

jeromekelleher commented:
OK, looking good. I'm going to leave this for a day or so in case someone wants to review, but I think we merge then.

jeromekelleher commented:
OK, we're good to go @daniel-goldstein. I made some tiny cosmetic tweaks and rebased. Will merge once CI passes. This is a huge jump forward, thanks!

gtsambos commented:
Yay what an exciting feature! Thanks for doing it @daniel-goldstein!

I'm just wondering, does this feature mean there's now a more elegant solution to the issue of simulating multiple chromosomes (documented [here](https://msprime.readthedocs.io/en/stable/tutorial.html#multiple-chromosomes) )?

Ie. does this mean that if I specify my recombination map with
```
positions = [chr1, chr2-1, chr2, chr3-1]
rates = [rho, 1, rho, 0]
```
that the resulting tree sequence will be of two independent chromosomes, with a tree breakpoint at coordinate `chr2`?

Asking because

 - I'm currently working on some new docs for msprime and can add this to the list of things to change
 - I'm currently designing some simulations that would work much better if I could do this...

jeromekelleher commented:
I don't think this will make any fundamental difference @gtsambos, as it's essentially the same model we're simulating as we've documented there. But, we will need to update the docs. I'll open an issue so we don't forget about it.

DomNelson commented:
I think it will be slightly cleaner though @gtsambos @jeromekelleher, since previously the hotspot added between chromosomes would end up with many distinct breakpoints/trees all jammed into it, since during the simulations it would be treated the same as any other region with the same recombination length. That meant you needed to be a bit careful iterating through all trees, since the hotspots would have a huge number of trees < 1 bp apart, and it also wasted a decent amount of simulation time generating those extra trees.

In physical coords all that mess is avoided, no? As I understand it you'd just pass directly from one chromosome to another, which does seem like an improvement. Also the rescaling to avoid bad edges (as described in the docs) would be unnecessary.

jeromekelleher commented:
Thinking about it again, the "phantom trees" thing shouldn't have existed anyway if we were doing the num_loci == sequence_length thing properly in the old code, so I don't know why that happened. I guess it should be fixed now anyway --- anyone want to try it out?

DomNelson commented:
Ran a quick check and things seem to be working for the Hudson but not DTWF - put things in a comment on #876

gtsambos commented:
Sorry for the late reply all, been a bit busy moving house.

> I think it will be slightly cleaner though @gtsambos @jeromekelleher, since previously the hotspot added between chromosomes would end up with many distinct breakpoints/trees all jammed into it,

Yes this is what I was thinking (and what I hoped would be fixed with this new feature!) I'm trying to simulate a chromosome with some intervals masked out -- it doesn't make sense to simulate the chunks independently since there's long-range LD, so I'm currently trying to approach shown in the current docs. You can set the space between the segments to be small enough that there are no (or negligibly many) mutations on them, but you still end up simulating a gazillion squished up trees in the space. As is , you still effectively have to simulate the tree topologies for the whole chromosome, which is v wasteful if you only care about certain small segments of the chromosome. But it looks like this feature does away with this (if I've understood correctly?)

jeromekelleher commented:
This should work cleanly now with ``discrete=True`` @gtsambos (although I still don't understand why we were getting lots of trees in the non-recombining bit if num_loci == sequence_length in the old code...)
